### ISSUE TITLE
State variable changes aren't monitored for custom class instances

### ISSUE THREAD

**Issue by ben-directai** [gradio-app/gradio#9221]
### Describe the bug

The [gradio state docs](https://www.gradio.app/docs/gradio/state) imply that any deepcopy-able object type is acceptable as an underlying value. There isn't documentation on how event listeners monitor changes to those variables. I'm experiencing unexpected behavior when attempting to monitor .change() events on instances of custom classes (inside of Gradio State objects).

Namely, modifying variables inside that custom class instance doesn't trigger its .change() event. Note that the logs demonstrate that the instance inside the custom State object is changing, but .change() never runs for that particular state variable.

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

```python
import gradio as gr

class SimpleClass:
    def __init__(self):
        self.numerical_val = 5

with gr.Blocks() as demo:
    simple_state = gr.State(SimpleClass())
    int_state = gr.State(5)
    
    num_slider = gr.Slider(minimum=0,maximum=10, value=5,step=1)
    
    def change_simple_state_num(simple_state_val, int_state_val, new_val):
        simple_state_val.numerical_val = new_val
        int_state_val = new_val
        print(simple_state_val.numerical_val)
        return simple_state_val, int_state_val
    
    num_slider.release(
        change_simple_state_num,
        inputs=[simple_state, int_state, num_slider],
        outputs=[simple_state, int_state]
    )
    
    simple_state.change(
        lambda x: print(f"The new value of simple_state.numerical_val is {x.numerical_val}"),
        simple_state
    )
    int_state.change(
        lambda x: print(f"The new value of int_state is {x}"),
        int_state
    )
    
demo.launch()

```

### Screenshot

<img width="1347" alt="Screenshot 2024-08-30 at 12 36 47 AM" src="https://github.com/user-attachments/assets/44c1d77c-73cb-4f34-b8e9-69023e884dc2">

### Logs

```shell
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
2
The new value of int_state is 2
9
The new value of int_state is 9
6
The new value of int_state is 6
10
The new value of int_state is 10
```
```


### System Info

```shell
gradio environment
Gradio Environment Information:
------------------------------
Operating System: Darwin
gradio version: 4.41.0
gradio_client version: 1.3.0

------------------------------------------------
gradio dependencies in your environment:

aiofiles: 23.2.1
anyio: 4.4.0
fastapi: 0.112.0
ffmpy: 0.4.0
gradio-client==1.3.0 is not installed.
httpx: 0.27.0
huggingface-hub: 0.24.5
importlib-resources: 6.4.2
jinja2: 3.1.4
markupsafe: 2.1.5
matplotlib: 3.9.2
numpy: 2.0.1
orjson: 3.10.7
packaging: 24.1
pandas: 2.2.2
pillow: 10.4.0
pydantic: 2.8.2
pydub: 0.25.1
python-multipart: 0.0.9
pyyaml: 6.0.2
ruff: 0.6.0
semantic-version: 2.10.0
tomlkit==0.12.0 is not installed.
typer: 0.12.3
typing-extensions: 4.12.2
urllib3: 2.2.2
uvicorn: 0.30.6
authlib; extra == 'oauth' is not installed.
itsdangerous; extra == 'oauth' is not installed.


gradio_client dependencies in your environment:

fsspec: 2024.6.1
httpx: 0.27.0
huggingface-hub: 0.24.5
packaging: 24.1
typing-extensions: 4.12.2
websockets: 12.0
```
```


### Severity

I can work around it

**Comment by freddyaboulton** [comment]
Hi @ben-directai - this is expected at the moment. You need to return a new instance of the class instead of modifying it in place. I think we can at least make the docs here clearer. Maybe change the implementation too? cc @aliabid94

**Comment by ben-directai** [comment]
thanks @freddyaboulton - that's a very helpful note

**Comment by abidlabs** [comment]
Closed in the `5.0-dev` branch, thanks @aliabid94!
